test/_mystery-exit.js

```diff
@@ -0,0 +1,111 @@
+var workerCount = require('./helpers').workerCount;
+
+cluster.setupMaster({
+  exec: 'test/workers/null.js'
+});
+
+tap.test('from size 1 to 0', function(t) {
+  master.start({size: 1});
+  master.once('startWorker', function() {
+    master.setSize(0);
+  });
+
+  master.once('stopWorker', function(worker) {
+    t.assert(worker.process.pid, 'worker is valid');
+    worker.on('exit', function() {
+      t.equal(workerCount(), 0);
+      master.stop(t.end);
+    });
+  });
+});
+
+tap.test('from size 3 to 1', function(t) {
+  master.start({size: 3});
+  master.on('startWorker', function() {
+    if (workerCount() === 3) {
+      master.setSize(1);
+    }
+  });
+  master.once('stopWorker', function() {
+    t.equal(workerCount(), 2);
+    master.once('stopWorker', function() {
+      t.equal(workerCount(), 1);
+      master.stop(t.end);
+    });
+  });
+  console.log('trying to stay alive');
+
+  process.nextTick(function() {
+    console.log('next');
+  });
+  setImmediate(function() {
+    console.log('immed');
+  });
+  setTimeout(function() {
+    console.log('timeout');
+  }, 3000);
+  var a = require('net').createServer().listen(0, '127.0.0.1');
+  var b = require('net').createServer().listen(0, '::');
+  var c = require('net').createServer().listen(9876, '127.0.0.1');
+
+  console.log(a && b && c);
+
+  var dump = {
+    handles: function() {
+      var fmt = require('util').format;
+      var inspect = require('util').inspect;
+      var handles = process._getActiveHandles();
+
+      console.log('ActiveHandles:');
+      handles.forEach(function(h, i) {
+        console.log('%d: %s', i, pretty(h));
+      });
+
+      function pretty(h) {
+        if (h === process.stderr)
+          return fmt('process.stderr/type=%s/fd=%j', h._type, h.fd);
+        if (h === process.stdout)
+          return fmt('process.stdout/type=%s/fd=%j', h._type, h.fd);
+        if (h === process.stdin)
+          return fmt('process.stdin/type=%s/fd=%j', h._type, h.fd);
+        if ('_idleNext' in h)
+          return fmt('timer:msecs=%s:count=%d', h.msecs, timerCount(h));
+        if ('_connections' in h)
+          return fmt('server:%j', h.address());
+        return inspect(h, {depth: 1});
+      }
+
+      function timerCount(h) {
+        // XXX the circular list always has one extra entry for each setTimeout
+        // call, why?
+        var count = -1;
+        var first = h._idleNext;
+        var next = first;
+        while (next) {
+          next = next._idleNext;
+          count++;
+          if (next === first)
+            break;
+        }
+        return count;
+      }
+    },
+    requests: function() {
+      console.log('ActiveRequests:', process._getActiveRequests());
+    },
+  };
+  dump.handles();
+  dump.requests();
+  process.on('exit', function(code) {
+    console.log('exit: code %d', code);
+  });
+  process.on('beforeExit', function(code) {
+    console.log('exit: code %d', code);
+  });
+  console.log('bye?');
```

Comment:
@bnoordhuis could you help me with this mystery, please? I am failing to understand why this script exits after logging "bye?". It has 3 servers listening, an immediate, a next tick, and a set timeout, and it exits, WITHOUT emitting the exit and beforeExit events, AFAICT. I'm not sure which is stranger, that the loop is dead, or that the events aren't emitted.

If I comment out the first tap test in the file, though, this second one runs to its expected completion. I'm not sure if I'm triggering a node bug, or if I'm just somehow damaging the node runtime (but how?).  I've even stepped through StartNodeInstance, and I can see the process emits, but nothing is printed. And uv definitely thinks its dead (according to uv__loop_alive).

Any ideas? Or even any ideas how to debug this? I'm at my wits end. For the purposes of the tests, I just split the tests up one per file, which is better test structure anyway, but its quite disconcerting to have node exiting like this - there is some kind of nasty bug somewhere, or I'm really misunderstanding something.

Output of above, FYI, on node 3.x (the handles active depends on node version, but all the versions I tested exit immediately despite the pending ticks, etc.).

```
sn/strong-cluster-control (fix-restart *%) % node test/_mystery-exit.js
TAP version 13
    # Subtest: from size 1 to 0
    ok 1 - worker is valid
    1..1
ok 1 - from size 1 to 0 # time=117.402ms

    # Subtest: from size 3 to 1
trying to stay alive
why do we die :-(
ActiveHandles:
0: process.stderr/type=tty/fd=2
1: process.stdout/type=tty/fd=1
2: timer:msecs=3000:count=1
ActiveRequests: []
bye?
```